### ISSUE TITLE
[Discussion] around optimizing delivery

### ISSUE THREAD

**Issue by itsjoekent** [DoSomething/phoenix-next#463]
[Google PageSpeed Insights](https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fdosomething.org%2Fus%2Fcampaigns%2Fsincerely-us&tab=mobile) currently rates this URL, `https://dosomething.org/us/campaigns/sincerely-us` as **Poor 0/100**. Personally, I don't think it's _that_ bad, but because Google says so, it has a large impact on our SEO & rankings. Based on best practices and suggestions from Google, I'd like to make the following changes to optimize delivery.

### react server side rendering

doing this is going to be a bit of a challenge, as it's possible some of our packages don't support it / conflict with it, i've also never done it before so there will be a learning curve. the upside is there will be content before the initial react render, which is critial for SEO, mobile performance and open graph crawlers that dont execute JS such as Facebooks.

rendering react server side with PHP requires a custom v8 plugin installed along with system config changes. nodejs however _is_ v8, so it can be done natively without any system changes or configuration. for this reason, I suggest splitting the P-N codebase in half, all frontend assets living in a new nodejs app which handles the server side render.

finally in doing server side render, I suggest we update to React 16. React 16 has [several critical improvements](https://reactjs.org/blog/2017/09/26/react-v16.0.html#better-server-side-rendering) for server side rendering.

### reduce asset size

there are a few tricks we can use to drastically cut down our file sizes. our main app file is as of right now is over 800KB on production, and the css file is 150KB.

**easy tasks**
1. implement gzip. this is very simple to do, it requires adding a plugin to our webpack config and a change to file serving logic.
2. minify css. we already have post-css when we build, but we don't appear to have minification turned on.
3. update to webpack 3. scope hoisting will destroy our build to pile of dust.

**harder tasks**
1. we should serve assets from a CDN, not the server public directory
2. we should implement headers to instruct browsers to use local cache for certain assets such as fonts, images 
3. once we start introducing more page types, such as static pages, we can make a different webpack entry point for them.

### optimize image usage

we're currently serving incredibly large, uncompressed and nonresized images straight from Contentful, which is seriosuly hurting our performance rankings. there are a few things we can do here to make this better, in order of difficulty

1. utilize the helper method to query the browser size, and pull in proper image cropping. the most important image to worry about here at the moment is the lede banner since thats above the fold. 
2. push our contentful images to [googles fancy image optimizer](https://github.com/google/guetzli), which will cut the image size by 20-30%. (just to be clear - this would happen server side, not client)
3. [support progressive image loading](https://jmperezperez.com/medium-image-progressive-loading-placeholder/). there are alternatives to this, eg: [Etsy](http://www.guypo.com/introducing-lqip-low-quality-image-placeholders/) and [Facebooks](https://code.facebook.com/posts/991252547593574/the-technology-behind-preview-photos/). I haven't spent much time looking into these, but I think its all a good starting point.

### optimize react render time

its still worth doing this even after we make the files smaller. in React 16 we can do this straight from the browser https://reactjs.org/docs/optimizing-performance.html#profiling-components-with-the-chrome-performance-tab

----

[copied from my agenda](https://gist.github.com/itsjoekent/59bb7de0b63f7618771dbc426d9e322a)

**Comment by itsjoekent** [comment]
one thing i just thought of with splitting the codebase, auth. with the oauth model I think it's well within the realm of possibility to solve this though, as many products have done an architecture exactly like this.

**Comment by DFurnes** [comment]
> rendering react server side with PHP requires a custom v8 plugin installed along with system config changes. nodejs however is v8, so it can be done natively without any system changes or configuration. for this reason, I suggest splitting the P-N codebase in half, all frontend assets living in a new nodejs app which handles the server side render.

Yeah, I really like the idea of a small front-end app for rendering the interface. We took a similar approach with a [separate "renderer" app](https://github.com/DoSomething/voting-app/blob/dev/resources/assets/js/server.js) in Voting App that we'd make internal API requests against (but cutting that out is even better!)

> finally in doing server side render, I suggest we update to React 16. React 16 has [several critical improvements](https://reactjs.org/blog/2017/09/26/react-v16.0.html#better-server-side-rendering) for server side rendering.

Yesssss, let's make sure we have a ticket for this in an upcoming sprint! I've heard there are some paint points with updating to the latest Enzyme, but otherwise should be totally painless!

> implement gzip. this is very simple to do, it requires adding a plugin to our webpack config and a change to file serving logic.

This is something we automatically do in Fastly elsewhere, we should do the same here.

> minify css. we already have post-css when we build, but we don't appear to have minification turned on.

Oy vey, we should definitely be doing this.

> we should serve assets from a CDN, not the server public directory

We do! Unless we explicitly disabled it, Fastly runs on all of our public traffic.

> [support progressive image loading](https://jmperezperez.com/medium-image-progressive-loading-placeholder/). there are alternatives to this, eg: [Etsy](http://www.guypo.com/introducing-lqip-low-quality-image-placeholders/) and [Facebooks](https://code.facebook.com/posts/991252547593574/the-technology-behind-preview-photos/). I haven't spent much time looking into these, but I think its all a good starting point.

Yeah, I'm very interested in doing this (at least for Rogue gallery responses, but could also be something that we could create a little proxy for doing with Contentful images as well!)

**Comment by DFurnes** [comment]
> one thing i just thought of with splitting the codebase, auth. with the oauth model I think it's well within the realm of possibility to solve this though, as many products have done an architecture exactly like this.

Yeah, that's the hope! There's maybe a bit more to think about when we can't just use the controllers packaged with Gateway, but I'm sure there are [good options](https://github.com/jaredhanson/passport-openidconnect) for this in Node as well.

**Comment by itsjoekent** [comment]
<img width="513" alt="screen shot 2017-10-20 at 2 55 28 pm" src="https://user-images.githubusercontent.com/897368/31837427-c55a3146-b5a6-11e7-9cc2-7ee677c3367a.png">

From P-A homepage, you'd appear to be right about that!
